File: apps/backend/app/schemas/jd.py
"""Pydantic schemas for JD-from-URL (P3 §D, Requirement 9)."""

from __future__ import annotations

from pydantic import BaseModel, Field


class FetchUrlRequest(BaseModel):
    """Fetch a job posting by URL for the tailor flow."""

    url: str = Field(min_length=1, max_length=2048)
    # Opt-in AI cleanup (R15 — never auto-fires; cost-aware).
    use_ai: bool = False


class ExtractRenderedRequest(BaseModel):
    """Browser-extension fallback: the user's already-rendered DOM (Phase 4)."""

    url: str = Field(min_length=1, max_length=2048)
    html: str = Field(min_length=1, max_length=5_242_880)  # 5 MiB cap
    use_ai: bool = False


class WebhookJobPayload(BaseModel):
    """Employer-pushed authoritative job posting (Phase 4, zero-scrape)."""

    url: str = Field(min_length=1, max_length=2048)
    title: str | None = Field(default=None, max_length=512)
    company: str | None = Field(default=None, max_length=512)
    location: str | None = Field(default=None, max_length=512)
    employment_type: str | None = Field(default=None, max_length=128)
    salary: str | None = Field(default=None, max_length=256)
    description: str | None = Field(default=None, max_length=131_072)
    description_html: str | None = Field(default=None, max_length=262_144)


class FetchUrlResponse(BaseModel):
    """Extracted JD; ``low_confidence`` prompts user verification before tailoring.

    v2 fields are optional and default to ``None`` so v1 responses are byte-for-byte
    unchanged. The frontend detects v2 by the presence of ``schema_version``.
    """

    content: str
    low_confidence: bool
    source_url: str

    # --- v2 metadata (optional, backward compatible) ---
    schema_version: str | None = None
    confidence_level: str | None = None      # HIGH / MEDIUM / LOW
    confidence_score: int | None = None
    source: str | None = None                # platform_api / json_ld / dom_semantic / ...
    partial: bool | None = None
    error_code: str | None = None
    language: str | None = None
    suggestions: list[str] | None = None
    warnings: list[str] | None = None

File: apps/frontend/lib/api/jd.ts
/**
 * JD-from-URL API (P3 §D, Requirement 9 + JD v2 enhancements).
 *
 * Posts a job URL to the SSRF-hardened backend fetcher and returns the extracted
 * description for the tailor flow. ``lowConfidence`` signals the UI to ask the
 * user to verify/edit the text before generating (never auto-tailors). Errors
 * are opaque by design (the backend never reveals *why* a fetch was blocked).
 *
 * The v2 cascade adds optional explainability metadata (``schemaVersion``,
 * ``confidenceLevel``, ``source``, ``suggestions``, ``warnings``, ``errorCode``).
 * These are all optional so a v1 backend response still parses unchanged; the UI
 * detects v2 by the presence of ``schemaVersion``.
 */
import { apiPost } from './client';

export type JdConfidence = 'HIGH' | 'MEDIUM' | 'LOW';

export interface JdFromUrl {
  content: string;
  lowConfidence: boolean;
  sourceUrl: string;
  // v2 (optional)
  schemaVersion?: string;
  confidenceLevel?: JdConfidence;
  confidenceScore?: number;
  source?: string;
  partial?: boolean;
  errorCode?: string;
  language?: string;
  suggestions?: string[];
  warnings?: string[];
}

interface RawJd {
  content: string;
  low_confidence: boolean;
  source_url: string;
  schema_version?: string | null;
  confidence_level?: JdConfidence | null;
  confidence_score?: number | null;
  source?: string | null;
  partial?: boolean | null;
  error_code?: string | null;
  language?: string | null;
  suggestions?: string[] | null;
  warnings?: string[] | null;
}

/** Fetch + extract a JD from a URL. ``useAi`` opts into bounded LLM cleanup. */
export async function fetchJdFromUrl(url: string, useAi = false): Promise<JdFromUrl> {
  const res = await apiPost('/jobs/fetch-url', { url, use_ai: useAi });
  if (res.status === 404) {
    throw new Error('Importing from a URL is currently unavailable.');
  }
  if (res.status === 429) {
    throw new Error('Too many imports right now. Please wait a moment and try again.');
  }
  if (!res.ok) {
    // Opaque backend failure - show a friendly, non-leaky message.
    throw new Error("We couldn't read that job posting. Paste the description instead.");
  }
  const body = (await res.json()) as RawJd;
  return {
    content: body.content,
    lowConfidence: body.low_confidence,
    sourceUrl: body.source_url,
    schemaVersion: body.schema_version ?? undefined,
    confidenceLevel: body.confidence_level ?? undefined,
    confidenceScore: body.confidence_score ?? undefined,
    source: body.source ?? undefined,
    partial: body.partial ?? undefined,
    errorCode: body.error_code ?? undefined,
    language: body.language ?? undefined,
    suggestions: body.suggestions ?? undefined,
    warnings: body.warnings ?? undefined,
  };
}

/** Human-readable label for the extraction source (for the "how" tooltip). */
export function jdSourceLabel(source?: string): string {
  switch (source) {
    case 'platform_api':
      return 'official job-board API';
    case 'json_ld':
      return 'structured page data';
    case 'hydration_json':
      return 'embedded page data';
    case 'dom_semantic':
      return 'page content';
    case 'headless_dom':
      return 'rendered browser page';
    case 'pdf_ocr':
      return 'PDF document';
    default:
      return 'the page';
  }
}
